server/src/main/java/org/apache/iotdb/db/qp/logical/crud/QueryOperator.java

```diff
@@ -337,6 +347,7 @@ protected AlignByDevicePlan generateAlignByDevicePlan(PhysicalGenerator generato
     alignByDevicePlan.setEnableTracing(enableTracing);
 
     alignByDevicePlan.deduplicate(generator);
+    checkWithoutNullColumnValidAlignByDevice(alignByDevicePlan);
```

Comment:
You fill `withoutNullColumns` of `AlignByDevicePlan` in `convertSpecialClauseValues` and fill `withoutNullValidSet` of  `AlignByDevicePlan` in `alignByDevicePlan.deduplicate(generator)`, then do this check whether all without null columns exit in withoutNullValidSet.

Actually, you can see that `withoutNullValidSet` in `AlignByDevicePlan` is unnecessary, you can make it as local variable here, and pass it as a parameter to `checkWithoutNullColumnValidAlignByDevice`.

BTW, add some comments about where you init `withoutNullColumns` of `AlignByDevicePlan` above this check method.

Reply:
fix